substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,581 @@
+ * We first analyze A and decide whether to enter the method for further analysis or not. In first
+ * case, we proceed and analyze the nodes deeper in the graph with MethodNodeTrackingAndInline. We
+ * repeat the procedure with B and C, and when the analysis of method C is complete, we decide to
+ * inline it because it only returns a constant value. After that, B folds to constant, so we inline
+ * this method too. Finally, for the same reason, we decide to inline A into R.
+ */
+public class NativeImageInlineDuringParsingPlugin implements InlineInvokePlugin {
+
+    public static class Options {
+        @Option(help = "Inline methods which folds to constant during parsing before the static analysis.")//
+        public static final HostedOptionKey<Boolean> InlineBeforeAnalysis = new HostedOptionKey<>(true);
+
+    }
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+
+    /**
+     * The map in which we place the result of a inlining decision. During analysis we store
+     * information about method in separate map using {@link #callSitemap} and {@link #add}. After
+     * analysis we read information from this map to find inline decision we made using
+     * {@link #get}.
+     */
+    private final ConcurrentHashMap<AnalysisMethod, ConcurrentHashMap<Pair<Integer, AnalysisMethod>, InvocationResult>> dataInline = new ConcurrentHashMap<>();
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    @Override
+    public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod callee, ValueNode[] args) {
+        if (b.parsingIntrinsic()) {
+            /* We are not interfering with any intrinsic method handling. */
+            return null;
+        }
+
+        if (callee.getAnnotation(NeverInline.class) != null || callee.getAnnotation(NeverInlineTrivial.class) != null || b.getMethod().getAnnotation(NeverInline.class) != null ||
+                        b.getMethod().getAnnotation(NeverInlineTrivial.class) != null) {
+            return null;
+        }
+
+        if (callee.getAnnotation(RestrictHeapAccess.class) != null || callee.getAnnotation(Uninterruptible.class) != null || b.getMethod().getAnnotation(RestrictHeapAccess.class) != null ||
+                        b.getMethod().getAnnotation(Uninterruptible.class) != null) {
+            /*
+             * Caller or callee have an annotation that might prevent inlining. We don't check the
+             * exact condition but instead always bail out for simplicity.
+             */
+            return null;
+        }
+
+        CallSite callSite = new CallSite(toAnalysisMethod(b.getMethod()), b.bci(), toAnalysisMethod(callee));
+        InvocationResult inline;
+
+        if (analysis) {
+            inline = ImageSingletons.lookup(NativeImageInlineDuringParsingPlugin.class).get(b.getMethod(), b.bci(), callee);
```

Comment:
The `ImageSingletons.lookup(NativeImageInlineDuringParsingPlugin.class)` pattern repets several times. A `NativeImageInlineDuringParsingPlugin.singleton()` method that returns the value would improve readability.

Reply:
Done, thanks for the suggestion.